File: Data Structures & Algorithms/valid-tree/submission-3.java
class Solution {
    boolean dfs(List<List<Integer>> adj, int u, boolean vis[], int par) {
        vis[u] = true;
        for(int v : adj.get(u)) {
            if(v == par) continue;
            if(vis[v] == true && par != u) return true;
            if(dfs(adj, v, vis, u) == true) return true;
        }
        return false;
    }
    public boolean validTree(int n, int[][] edges) {
        int r = edges.length;
        if(r < n-1) return false;
        boolean vis[] = new boolean[n];
        int parent = -1;

        List<List<Integer>> adj = new ArrayList<>();
        for(int i = 0; i < n; i++) adj.add(new ArrayList<>());
        for(int[] edge : edges) {
            int u = edge[0], v = edge[1];
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        for(int i = 0; i < n; i++) {
           if(!vis[i] && dfs(adj, i, vis, parent)) {
            return false;
           }
        }
        for(boolean i : vis) {
            if(i == false) return false;
        }
        return true;
    }
}
/*
3
|
0-1-4
|
2

0-1-2-3

*/

File: Data Structures & Algorithms/count-number-of-islands/submission-3.cpp
class Solution {
public:
    bool dfs(vector<vector<char>>& grid, int n, int m, 
            vector<vector<bool>>& vis, int i, int j) {
                if(i < 0 || i >= n || j >= m || j < 0 || vis[i][j] == true
                ||grid[i][j] == '0') return false;
                
                vis[i][j] = true;
                //if(i+1 < n && grid[i+1][j] == '1')
                    dfs(grid, n, m, vis, i+1, j); 
                //if(j-1 >= 0 && grid[i][j-1] == '1')
                    dfs(grid, n, m, vis, i, j-1);
                //if(j+1 < n && grid[i][j+1] == '1')
                    dfs(grid, n, m, vis, i, j+1);
                //if(i-1 >= 0 && grid[i-1][j] == '1')
                    dfs(grid, n, m, vis, i-1, j);
                return true;
            }
    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size(), m = grid[0].size(), island = 0;
        vector<vector<bool>> vis(n, vector<bool>(m, false));
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(grid[i][j] == '1') {
                    //cout << i <<": "<<j<<endl;
                    if(dfs(grid, n, m, vis, i, j))
                    island++;
                }
            }
        }
        
        return island;
    }
};
